### DISCUSSION TITLE
Question about vips_text(), loading font from font file.

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2265] [Q&A] [answered]
Hello, I have a question about vips_text().
I use vips_text() like this
```
text_image = vips::VImage::text(text.c_str(),
                vips::VImage::option ()->
                set("font", font.c_str()) ->
                set("fontfile", text_fontfile.c_str()) ->
                set("align",  VIPS_ALIGN_CENTRE) ->
                set("width",  text_image_width) ->
                set("spacing", (text_lineHeight - 1) * text_fontsize));
```
When I parallel this function, sometimes it takes a long time.
If the font files are the same, will the the font file be locked when this function is parallel?

**Comment by jcupitt** [comment]
Hi, yes, `text` is single-threaded because on many platforms the font system is single-threaded (especially loading new fonts).

**Reply by rebornfz** [reply]
@jcupitt 
Is there any difference between setting the fontfile parameter and not setting the fontfile parameter?
If the fontfile parameter is not set, is the font file loaded from the system as well?

**Reply by jcupitt** [reply]
The `fontfile` parameter adds the named font file to the list of fonts known to the font system. If the font is already installed, you don't need to use it.

**Comment by jcupitt** [comment, marked as answer]
I checked the sources to remind myself, and the problem is `PangoFontMap`. There's a helpful comment:

```C
/* Just have one of these and reuse it.
 *
 * This does not unref cleanly on many platforms, so we will leak horribly
 * unless we reuse it. Sadly this means vips_text() needs to use a lock 
 * internally to single-thread text rendering.
 */
static PangoFontMap *vips_text_fontmap = NULL;
```

So `text` is single-threaded because all vips text rendering uses a single shared fontmap, and there's a single shared fontmap because many font rendering backends leak horribly.
